src/darwin/CHIPTool/CHIPTool/View Controllers/Custom Flow/CustomFlowViewController.m

```diff
@@ -0,0 +1,586 @@
+    _activityIndicator.hidden = YES;
+    _errorLabel.hidden = YES;
+}
+
+- (void)qrCodeInitialState
+{
+    if ([_captureSession isRunning]) {
+        [_captureSession stopRunning];
+    }
+    if ([_activityIndicator isAnimating]) {
+        [_activityIndicator stopAnimating];
+    }
+    [self changeNavBarButtonToCamera];
+    _activityIndicator.hidden = YES;
+    _captureSession = nil;
+    [_videoPreviewLayer removeFromSuperlayer];
+}
+
+- (void)scanningStartState
+{
+    [self changeNavBarButtonToCancel];
+    _setupPayloadView.hidden = YES;
+    _deviceModelInfoView.hidden = YES;
+    _readFromLedgerButton.hidden = YES;
+    _redirectButton.hidden = YES;
+    _errorLabel.hidden = YES;
+}
+
+- (void)postScanningQRCodeState
+{
+    _captureSession = nil;
+    [self changeNavBarButtonToCamera];
+
+    [_videoPreviewLayer removeFromSuperlayer];
+
+    self->_activityIndicator.hidden = NO;
+    [self->_activityIndicator startAnimating];
+}
+
+- (void)showError:(NSError *)error
+{
+    [self->_activityIndicator stopAnimating];
+    self->_activityIndicator.hidden = YES;
+    _readFromLedgerButton.hidden = YES;
+    _redirectButton.hidden = YES;
+
+    self->_errorLabel.text = error.localizedDescription;
+    self->_errorLabel.hidden = NO;
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, ERROR_DISPLAY_TIME), dispatch_get_main_queue(), ^{
+        self->_errorLabel.hidden = YES;
+    });
+}
+
+- (void)showPayload:(CHIPSetupPayload *)payload rawPayload:(NSString *)rawPayload
+{
+    [self->_activityIndicator stopAnimating];
+    self->_activityIndicator.hidden = YES;
+    self->_errorLabel.hidden = YES;
+    // reset the view and remove any preferences that were stored from a previous scan
+    self->_setupPayloadView.hidden = NO;
+
+    [self updateUIFields:payload rawPayload:rawPayload];
+}
+
+- (void)updateUIFields:(CHIPSetupPayload *)payload rawPayload:(nullable NSString *)rawPayload
+{
+    _versionLabel.text = [NSString stringWithFormat:@"%@", payload.version];
+    _rendezVousInformation.text = [NSString stringWithFormat:@"%lu", payload.rendezvousInformation];
+    if ([payload.serialNumber length] > 0) {
+        self->_serialNumber.text = payload.serialNumber;
+    } else {
+        self->_serialNumber.text = NOT_APPLICABLE_STRING;
+    }
+
+    _discriminatorLabel.text = [NSString stringWithFormat:@"%@", payload.discriminator];
+    _setupPinCodeLabel.text = [NSString stringWithFormat:@"%@", payload.setUpPINCode];
+    // TODO: Only display vid and pid if present
+    _vendorID.text = [NSString stringWithFormat:@"%@", payload.vendorID];
+    _productID.text = [NSString stringWithFormat:@"%@", payload.productID];
+    _vendorIDDeviceInfo.text = [NSString stringWithFormat:@"%@", payload.vendorID];
+    _productIDDeviceInfo.text = [NSString stringWithFormat:@"%@", payload.productID];
+    _commissioningFlow.text = [NSString stringWithFormat:@"%lu", payload.commissioningFlow];
+    
+    if (payload.commissioningFlow == 2) {
```

Comment:
Use of Enum ensure better reusability, for example CHIPCommissioningFlow.kCommissioningFlowCustom

Reply:
Yes.